ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,90 +27,253 @@
  */
 public class BlobId extends StoreKey {
-  private Short version = 1;
-  private static short Version_Size_In_Bytes = 2;
+  public static final byte DEFAULT_FLAG = 0;
+  // version 1 of the serialized format
+  static final short BLOB_ID_V1 = 1;
+  // version 2 of the serialized format
+  static final short BLOB_ID_V2 = 2;
+  private static final short CURRENT_VERSION = BLOB_ID_V1;
+  private static final short VERSION_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  private static final short UUID_SIZE_FIELD_LENGTH_IN_BYTES = Integer.BYTES;
+  private static final short FLAG_FIELD_LENGTH_IN_BYTES = Byte.BYTES;
+  private static final short DATACENTER_ID_FIELD_LENGTH_IN_BYTES = Byte.BYTES;
+  private static final short ACCOUNT_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  private static final short CONTAINER_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  // the version to indicate the serialized format.
+  private Short version;
+  private Byte flag;
+  private Byte datacenterId;
+  private Short accountId;
+  private Short containerId;
   private PartitionId partitionId;
   private String uuid;
-  private static int UUID_Size_In_Bytes = 4;
 
   /**
-   * Constructs a new unique BlobId for the specified partition.
-   *
-   * @param partitionId of Partition in which blob is to be stored.
+   * Constructs a new BlobId by taking arguments for the required fields. The constructed BlobId will be serialized
+   * into {@link #CURRENT_VERSION}. If {@code CURRENT_VERSION == BLOB_ID_V1}, it will serialize itself into
+   * {@code BLOB_ID_V1}, ignoring {@code flag}, {@code datacenterId}, {@code containerId}, and {@code containerId}
+   * regardless they are set or not.
+   * @param flag A byte to embed additional information of this blobId. Will be reset to {@link #DEFAULT_FLAG} if
+   *             {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param datacenterId The id of the datacenter to be embedded into the blob. Will be reset to
+   *             {@link ClusterMapUtils#LEGACY_DATACENTER_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param accountId The id of the {@link Account} to be embedded into the blob. Will be reset to
+   *             {@link Account#LEGACY_ACCOUNT_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param containerId The id of the {@link Container} to be embedded into the blob. Will be reset to
+   *             {@link Container#LEGACY_CONTAINER_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param partitionId The partition where this blob is to be stored. Cannot be {@code null}.
    */
-  public BlobId(PartitionId partitionId) {
+  public BlobId(byte flag, byte datacenterId, short accountId, short containerId, PartitionId partitionId) {
     if (partitionId == null) {
-      throw new IllegalArgumentException("Partition ID cannot be null");
+      throw new IllegalArgumentException("partitionId cannot be null");
+    }
+    version = getCurrentVersion();
+    switch (version) {
```

Comment:
It depends on how to treat the passed-in values if they are not required. For example, if I build a `blobId` with `flag=3`, but `CURRENT_VERSION` is 1, then what should we expect when we call `blobId.getFlag()`. We can either return 3, or `DEFAULT_FLAG`. No one is absolutely correct or wrong, but I am inclined to return the latter, because this will provide consistent result after serdes.